<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>New and modified functionality in xcms • xcms</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="New and modified functionality in xcms">
<meta property="og:description" content="xcms">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">xcms</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">3.11.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/xcms.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/new_functionality.html">New and modified functionality in xcms</a>
    </li>
    <li>
      <a href="../articles/xcms-direct-injection.html">Grouping FTICR-MS data with xcms</a>
    </li>
    <li>
      <a href="../articles/xcms-lcms-ms.html">LC-MS/MS data analysis with xcms</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sneumann/xcms/">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>New and modified functionality in xcms</h1>
                        <h4 class="author">Johannes Rainer</h4>
                        
      
      <small class="dont-index">Source: <a href="https://github.com/sneumann/xcms/blob/master/vignettes/new_functionality.Rmd"><code>vignettes/new_functionality.Rmd</code></a></small>
      <div class="hidden name"><code>new_functionality.Rmd</code></div>

    </div>

    
    
<script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  document.querySelector("h1").className = "title";
});
</script><script type="text/javascript">
document.addEventListener("DOMContentLoaded", function() {
  var links = document.links;  
  for (var i = 0, linksLength = links.length; i < linksLength; i++)
    if (links[i].hostname != window.location.hostname)
      links[i].target = '_blank';
});
</script><div id="new-functionality-in-xcms" class="section level1">
<h1 class="hasAnchor">
<a href="#new-functionality-in-xcms" class="anchor"></a>New functionality in <code>xcms</code>
</h1>
<p>This document describes new functionality and changes to existing functionality in the <code>xcms</code> package introduced during the update to version <em>3</em>.</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">xcms</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">RColorBrewer</span>)
<span class="fu">register</span>(<span class="fu">SerialParam</span>())</pre></body></html></div>
<div id="modernized-user-interface" class="section level2">
<h2 class="hasAnchor">
<a href="#modernized-user-interface" class="anchor"></a>Modernized user interface</h2>
<p>The modernization of the user interface comprises new classes for data representation and new data analysis methods. In addition, the core logic for the data processing has been extracted from the old methods and put into a set of R functions, the so called core API functions (or <code>do_</code> functions). These functions take standard R data structures as input and return standard R data types as result and can hence be easily included in other R packages.</p>
<p>The new user interface aims at simplifying and streamlining the <code>xcms</code> workflow while guaranteeing data integrity and performance also for large scale metabolomics experiments. Importantly, a simplified access to the original raw data should be provided throughout the whole metabolomics data analysis workflow.</p>
<p>The new interface re-uses objects from the <code>MSnbase</code> Bioconductor package, such as the <code>OnDiskMSnExp</code> object. This object is specifically designed for large scale MS experiments as it initially reads just the scan header information from the mzML while the mz-intensity value pairs from all or from selected spectra of a file are read on demand hence minimizing the memory demand. Also, in contrast to the old <code>xcmsRaw</code> object, the <code>OnDiskMSnExp</code> contains information from all files of an experiment. In addition, all data normalization and adjustment methods implemented in the <code>MSnbase</code> package can be directly applied to the MS data without the need to re-implement such methods in <code>xcms</code>. Results from <code>xcms</code> preprocessings, such as chromatographic peak detection or correspondence are stored into the new <code>XCMSnExp</code> object. This object extends the <code>OnDiskMSnExp</code> object and inherits thus all of its methods including raw data access.</p>
<p>Class and method/function names follow also a new naming convention trying tp avoid the partially confusing nomenclature of the original <code>xcms</code> methods (such as the <code>group</code> method to perform the correspondence of peaks across samples). To distinguish them from mass peaks, the peaks identified by the peak detection in an LS/GC-MS experiment are referred to as <em>chromatographic peaks</em>. The respective method to identify such peaks is hence called <code>findChromPeaks</code> and the identified peaks can be accessed using the <code>XCMSnExp</code> <code>chromPeaks</code> method. The results from an correspondence analysis which aims to match and group chromatographic peaks within and between samples are called <em>features</em>. A feature corresponds to individual ions with a unique mass-to-charge ratio (mz) and a unique retention time (rt). The definition of such mz-rt features (i.e. the result from the <code>groupChromPeaks</code> method) can be accessed <em>via</em> the <code>featureDefinitions</code> method of the <code>XCMSnExp</code> class. Finally, alignment (retention time correction) can be performed using the <code>adjustRtime</code> method.</p>
<p>The settings for any of the new analysis methods are bundled in <em>parameter</em> classes, one class for each method. This encapsulation of the parameters to a function into a parameter class (such as <code>CentWaveParam</code>) avoids busy function calls (with many single parameters) and enables saving, reloading and reusing the settings. In addition, the parameter classes are added, along with other information to the process history of an <code>XCMSnExp</code> object thus providing a detailed documentation of each processing step of an analysis, with the possibility to recall all settings of the performed analyses at any stage. In addition, validation of the parameters can be performed within the parameter object and hence is no longer required in the analysis function.</p>
</div>
<div id="new-naming-convention" class="section level2">
<h2 class="hasAnchor">
<a href="#new-naming-convention" class="anchor"></a>New naming convention</h2>
<p>Peaks identified in LC/GC-MS metabolomics are referred to as <em>chromatographic peaks</em> where possible to avoid any misconceptions with <em>mass peaks</em> identified in mz dimension.</p>
<p>Methods for data analysis from the original <code>xcms</code> code have been renamed to avoid potential confusions:</p>
<ul>
<li><p><strong>Chromatographic peak detection</strong>: <code>findChromPeaks</code> instead of <code>findPeaks</code>: for new functions and methods the term <em>peak</em> is avoided as much as possible, as it is usually used to describe a mass peak in mz dimension. To clearly distinguish between these peaks and peaks in retention time space, the latter are referred to as <em>chromatographic peak</em>, or <code>chromPeak</code>.</p></li>
<li><p><strong>Correspondence</strong>: <code>groupChromPeaks</code> instead of <code>group</code> to clearly indicate what is being grouped. Group might be a sample group or a peak group, the latter being referred to also by (mz-rt) <em>feature</em>.</p></li>
<li><p><strong>Alignment</strong>: <code>adjustRtime</code> instead of <code>retcor</code> for retention time correction. The word <em>cor</em> in <em>retcor</em> might be easily misinterpreted as <em>correlation</em> instead of correction.</p></li>
</ul>
</div>
<div id="new-data-classes" class="section level2">
<h2 class="hasAnchor">
<a href="#new-data-classes" class="anchor"></a>New data classes</h2>
<div id="ondiskmsnexp" class="section level3">
<h3 class="hasAnchor">
<a href="#ondiskmsnexp" class="anchor"></a><code>OnDiskMSnExp</code>
</h3>
<p>This object is defined and documented in the <code>MSnbase</code> package. In brief, it is a container for the full raw data from an MS-based experiment. To keep the memory footprint low the mz and intensity values are only loaded from the raw data files when required. The <code>OnDiskMSnExp</code> object replaces the <code>xcmsRaw</code> object.</p>
</div>
<div id="xcmsnexp" class="section level3">
<h3 class="hasAnchor">
<a href="#xcmsnexp" class="anchor"></a><code>XCMSnExp</code>
</h3>
<p>The <code>XCMSnExp</code> class extends the <code>OnDiskMSnExp</code> object from the <code>MSnbase</code> package and represents a container for the xcms-based preprocessing results while (since it inherits all functionality from its parent class) keeping a direct relation to the (raw) data on which the processing was performed. An additional slot <code>.processHistory</code> in the object allows to keep track of all performed processing steps. Each analysis method, such as <code>findChromPeaks</code> adds an <code>XProcessHistory</code> object which includes also the parameter class passed to the analysis method. Hence not only the time and type of the analysis, but its exact settings are reported within the <code>XCMSnExp</code> object. The <code>XCMSnExp</code> is thus equivalent to the <code>xcmsSet</code> from the original <code>xcms</code> implementation, but keeps in addition a link to the raw data on which the preprocessing was performed.</p>
</div>
<div id="chromatogram" class="section level3">
<h3 class="hasAnchor">
<a href="#chromatogram" class="anchor"></a><code>Chromatogram</code>
</h3>
<p>The <code>Chromatogram</code> class (available in the <code>MSnbase</code> package since version 2.3.8) allows a data representation that is orthogonal to the <code>Spectrum</code> class (also defined in <code>MSnbase</code>). The <code>Chromatogram</code> class stores retention time and intensity duplets and is designed to accommodate most use cases, from total ion chromatogram, base peak chromatogram to extracted ion chromatogram and SRM/MRM ion traces.</p>
<p><code>Chromatogram</code> objects can be extracted from <code>XCMSnExp</code> (and <code>MSnExp</code> and <code>OnDiskMSnExp</code>) objects using the <code>chromatogram</code> method.</p>
<p>Note that this class is still considered developmental and might thus undergo some changes in the future.</p>
</div>
</div>
<div id="binning-and-missing-value-imputation-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#binning-and-missing-value-imputation-functions" class="anchor"></a>Binning and missing value imputation functions</h2>
<p>The binning/profile matrix generation functions have been completely rewritten. The new <code>binYonX</code> function replaces the binning of intensity values into bins defined by their m/z values implemented in the <code>profBin</code>, <code>profBinLin</code> and <code>profBinLinBase</code> methods. The <code>binYonX</code> function provides also additional functionality:</p>
<ul>
<li><p>Breaks for the bins can be defined based on either the number of desired bins (<code>nBins</code>) or the size of a bin (<code>binSize</code>). In addition it is possible to provide a vector with pre-defined breaks. This allows to bin data from multiple files or scans on the same bin-definition.</p></li>
<li><p>The function returns a list with element <code>y</code> containing the binned values and element <code>x</code> the bin mid-points.</p></li>
<li><p>Values in input vector <code>y</code> can be aggregated within each bin with different methods: <code>max</code>, <code>min</code>, <code>sum</code> and <code>mean</code>.</p></li>
<li><p>The index of the largest (or smallest for <code>method</code> being “min”) within each bin can be returned by setting argument <code>returnIndex</code> to <code>TRUE</code>.</p></li>
<li><p>Binning can be performed on single or multiple sub-sets of the input vectors using the <code>fromIdx</code> and <code>toIdx</code> arguments. This replaces the <em>M</em> methods (such as <code>profBinM</code>). These sub-sets can be overlapping.</p></li>
</ul>
<p>The missing value imputation logic inherently build into the <code>profBinLin</code> and <code>profBinLinBase</code> methods has been implemented in the <code>imputeLinInterpol</code> function.</p>
<p>The example below illustrates the binning and imputation with the <code>binYtoX</code> and <code>imputeLinInterpol</code> functions. After binning of the test vectors below some of the bins have missing values, for which we impute a value using <code>imputeLinInterpol</code>. By default, <code>binYonX</code> selects the largest value within each bin, but other aggregation methods are also available (i.e. min, max, mean, sum).</p>
<div class="sourceCode" id="cb2"><html><body><pre class="r"><span class="co">## Defining the variables:</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">123</span>)
<span class="no">X</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="fl">30</span>, <span class="kw">mean</span> <span class="kw">=</span> <span class="fl">20</span>, <span class="kw">sd</span> <span class="kw">=</span> <span class="fl">25</span>))) <span class="co">## 10</span>
<span class="no">Y</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="fl">30</span>, <span class="kw">mean</span> <span class="kw">=</span> <span class="fl">50</span>, <span class="kw">sd</span> <span class="kw">=</span> <span class="fl">30</span>))

<span class="co">## Bin the values in Y into 20 bins defined on X</span>
<span class="no">res</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/binYonX.html">binYonX</a></span>(<span class="no">X</span>, <span class="no">Y</span>, <span class="kw">nBins</span> <span class="kw">=</span> <span class="fl">22</span>)

<span class="no">res</span></pre></body></html></div>
<pre><code>## $x
##  [1]  3.207154  6.066022  8.924891 11.783759 14.642628 17.501497 20.360365
##  [8] 23.219234 26.078102 28.936971 31.795840 34.654708 37.513577 40.372445
## [15] 43.231314 46.090183 48.949051 51.807920 54.666788 57.525657 60.384526
## [22] 63.243394
## 
## $y
##  [1]  76.85377  76.34400  48.14265  29.15879  43.76248        NA 115.06868
##  [8]  86.23886        NA  73.39895  49.14360        NA  91.05807  43.22687
## [15]        NA        NA        NA  95.49412        NA        NA  67.53841
## [22]  56.47825</code></pre>
<p>As a result we get a <code>list</code> with the bin mid-points (<code>$x</code>) and the binned <code>y</code> values (<code>$y</code>).</p>
<p>Next we use two different imputation approaches, a simple linear interpolation and the linear imputation approach that was defined in the <code>profBinLinBase</code> method. The latter performs linear interpolation only considering a certain neighborhood of missing values otherwise replacing the <code>NA</code> with a base value.</p>
<div class="sourceCode" id="cb4"><html><body><pre class="r"><span class="co">## Plot the actual data values.</span>
<span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="no">X</span>, <span class="no">Y</span>, <span class="kw">pch</span> <span class="kw">=</span> <span class="fl">16</span>, <span class="kw">ylim</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">Y</span>)))
<span class="co">## Visualizing the bins</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span>(<span class="kw">v</span> <span class="kw">=</span> <span class="fu"><a href="../reference/breaks_on_nBins.html">breaks_on_nBins</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span>(<span class="no">X</span>), <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">X</span>), <span class="kw">nBins</span> <span class="kw">=</span> <span class="fl">22</span>), <span class="kw">col</span> <span class="kw">=</span> <span class="st">"grey"</span>)

<span class="co">## Define colors:</span>
<span class="no">point_colors</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="fu"><a href="https://rdrr.io/pkg/RColorBrewer/man/ColorBrewer.html">brewer.pal</a></span>(<span class="fl">4</span>, <span class="st">"Set1"</span>), <span class="fl">80</span>)
<span class="co">## Plot the binned values.</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">res</span>$<span class="no">x</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">res</span>$<span class="no">y</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">point_colors</span>[<span class="fl">1</span>], <span class="kw">pch</span> <span class="kw">=</span> <span class="fl">15</span>)

<span class="co">## Perform the linear imputation.</span>
<span class="no">res_lin</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/imputeLinInterpol.html">imputeLinInterpol</a></span>(<span class="no">res</span>$<span class="no">y</span>)

<span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">res</span>$<span class="no">x</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">res_lin</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">point_colors</span>[<span class="fl">2</span>], <span class="kw">type</span> <span class="kw">=</span> <span class="st">"b"</span>)

<span class="co">## Perform the linear imputation "linbase"</span>
<span class="no">res_linbase</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/imputeLinInterpol.html">imputeLinInterpol</a></span>(<span class="no">res</span>$<span class="no">y</span>, <span class="kw">method</span> <span class="kw">=</span> <span class="st">"linbase"</span>)
<span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">res</span>$<span class="no">x</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">res_linbase</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="no">point_colors</span>[<span class="fl">3</span>], <span class="kw">type</span> <span class="kw">=</span> <span class="st">"b"</span>, <span class="kw">lty</span> <span class="kw">=</span> <span class="fl">2</span>)</pre></body></html></div>
<div class="figure">
<img src="new_functionality_files/figure-html/binning-imputation-example-1.png" alt="Binning and missing value imputation results. Black points represent the input values, red the results from the binning and blue and green the results from the imputation (with method lin and linbase, respectively)." width="960"><p class="caption">
Binning and missing value imputation results. Black points represent the input values, red the results from the binning and blue and green the results from the imputation (with method lin and linbase, respectively).
</p>
</div>
<p>The difference between the linear interpolation method <code>lin</code> and <code>linbase</code> is that the latter only performs the linear interpolation in a pre-defined neighborhood of the bin with the missing value (<code>1</code> by default). The other missing values are set to a base value corresponding to half of the smallest bin value. Both methods thus yield same results, except for bins 15-17 (see Figure above).</p>
</div>
<div id="core-functionality-exposed-via-simple-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#core-functionality-exposed-via-simple-functions" class="anchor"></a>Core functionality exposed <em>via</em> simple functions</h2>
<p>The core logic from the chromatographic peak detection methods <code>findPeaks.centWave</code>, <code>findPeaks.massifquant</code>, <code>findPeaks.matchedFilter</code> and <code>findPeaks.MSW</code> and from all alignment (<code>group.*</code>) and correspondence (<code>retcor.*</code>) methods has been extracted and put into functions with the common prefix <code>do_findChromPeaks</code>, <code>do_adjustRtime</code> and <code>do_groupChromPeaks</code>, respectively, with the aim, as detailed in issue <a href="https://github.com/sneumann/xcms/issues/30">#30</a>, to separate the core logic from the analysis methods invoked by the users to enable also the use these methods using base R parameters (i.e. without specific classes containing the data such as the <code>xcmsRaw</code> class). This simplifies also the re-use of these functions in other packages and simplifies the future implementation of the peak detection algorithms for e.g. the <code>MSnExp</code> or <code>OnDiskMSnExp</code> objects from the <code>MSnbase</code> Bioconductor package. The implemented functions are:</p>
<ul>
<li>
<strong>peak detection methods</strong>:
<ul>
<li>
<code>do_findChromPeaks_centWave</code>: peak density and wavelet based peak detection for high resolution LC/MS data in centroid mode <span class="citation">[1]</span>.</li>
<li>
<code>do_findChromPeaks_matchedFilter</code>: identification of peak in the chromatographic domain based on matched filtration <span class="citation">[2]</span>.</li>
<li>
<code>do_findChromPeaks_massifquant</code>: identification of peaks using Kalman filters.</li>
<li>
<code>do_findChromPeaks_MSW</code>: single spectrum, non-chromatographic peak detection.</li>
</ul>
</li>
<li>
<strong>alignment methods</strong>:
<ul>
<li>
<code>do_adjustRtime_peakGroups</code>: perform sample alignment (retention time correction) using alignment of <em>well behaved</em> chromatographic peaks that are present in most samples (and are expected to have the same retention time).</li>
</ul>
</li>
<li>
<strong>correspondence methods</strong>:
<ul>
<li>
<code>do_groupChromPeaks_density</code>: perform chromatographic peak grouping (within and across samples) based on the density distribution of peaks along the retention time axis.</li>
<li>
<code>do_groupChromPeaks_nearest</code>: groups peaks across samples similar to the method implemented in mzMine.</li>
<li>
<code>do_groupChromPeaks_mzClust</code>: performs high resolution correspondence on single spectra samples.</li>
</ul>
</li>
</ul>
<p>One possible drawback from the introduction of this new layer is, that more objects get copied by R which <em>could</em> eventually result in a larger memory demand or performance decrease (while no such was decrease was observed up to now).</p>
</div>
<div id="usability-improvements-in-the-old-user-interface" class="section level2">
<h2 class="hasAnchor">
<a href="#usability-improvements-in-the-old-user-interface" class="anchor"></a>Usability improvements in the <em>old</em> user interface</h2>
<ul>
<li>
<code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> subsetting method for <code>xcmsRaw</code> objects that enables to subset an <code>xcmsRaw</code> object to specific scans/spectra.</li>
<li>
<code>profMat</code> method to extract the <em>profile</em> matrix from the <code>xcmsRaw</code> object. This method should be used instead of directly accessing the <code>@env$profile</code> slot, as it will create the profile matrix on the fly if it was not pre-calculated (or if profile matrix generation settings have been changed).</li>
</ul>
</div>
</div>
<div id="changes-due-to-bug-fixes-and-modified-functionality" class="section level1">
<h1 class="hasAnchor">
<a href="#changes-due-to-bug-fixes-and-modified-functionality" class="anchor"></a>Changes due to bug fixes and modified functionality</h1>
<div id="differences-in-linear-interpolation-of-missing-values-profbinlin-" class="section level2">
<h2 class="hasAnchor">
<a href="#differences-in-linear-interpolation-of-missing-values-profbinlin-" class="anchor"></a>Differences in linear interpolation of missing values (<code>profBinLin</code>).</h2>
<p>From <code>xcms</code> version 1.51.1 on the new binning functions are used, thus, the bug described here are fixed.</p>
<p>Two bugs are present in the <code>profBinLin</code> method (reported as issues <a href="https://github.com/sneumann/xcms/issues/46">#46</a> and <a href="https://github.com/sneumann/xcms/issues/49">#49</a> on github) which are fixed in the new <code>binYonX</code> and <code>imputeLinInterpol</code> functions:</p>
<ul>
<li>The first bin value calculated by <code>profBinLin</code> can be wrong (i.e. not being the max value within that bin, but the first).</li>
<li>If the last bin contains also missing values, the method fails to determine a correct value for that bin.</li>
</ul>
<p>The <code>profBinLin</code> method is used in <code>findPeaks.matchedFilter</code> if the profile method is set to “binlin”.</p>
<p>The example below illustrates both differences.</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="co">## Define a vector with empty values at the end.</span>
<span class="no">X</span> <span class="kw">&lt;-</span> <span class="fl">1</span>:<span class="fl">11</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span>(<span class="fl">123</span>)
<span class="no">Y</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sort.html">sort</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="fl">11</span>, <span class="kw">mean</span> <span class="kw">=</span> <span class="fl">20</span>, <span class="kw">sd</span> <span class="kw">=</span> <span class="fl">10</span>))
<span class="no">Y</span>[<span class="fl">9</span>:<span class="fl">11</span>] <span class="kw">&lt;-</span> <span class="fl">NA</span>
<span class="no">nas</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span>(<span class="no">Y</span>)
<span class="co">## Do interpolation with profBinLin:</span>
<span class="no">resX</span> <span class="kw">&lt;-</span> <span class="kw pkg">xcms</span><span class="kw ns">:::</span><span class="fu"><a href="../reference/profGenerate.html">profBinLin</a></span>(<span class="no">X</span>[!<span class="no">nas</span>], <span class="no">Y</span>[!<span class="no">nas</span>], <span class="fl">5</span>, <span class="kw">xstart</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span>(<span class="no">X</span>),
                          <span class="kw">xend</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">X</span>))</pre></body></html></div>
<pre><code>## Warning: Use of 'profBinLin' is deprecated! Use 'binYonX' with
## 'imputeLinInterpol' instead.</code></pre>
<div class="sourceCode" id="cb7"><html><body><pre class="r"><span class="no">resX</span></pre></body></html></div>
<pre><code>## [1]  7.349388 15.543380 21.292877  0.000000  0.000000</code></pre>
<div class="sourceCode" id="cb9"><html><body><pre class="r"><span class="no">res</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/binYonX.html">binYonX</a></span>(<span class="no">X</span>, <span class="no">Y</span>, <span class="kw">nBins</span> <span class="kw">=</span> <span class="fl">5L</span>, <span class="kw">shiftByHalfBinSize</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="no">resM</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/imputeLinInterpol.html">imputeLinInterpol</a></span>(<span class="no">res</span>$<span class="no">y</span>, <span class="kw">method</span> <span class="kw">=</span> <span class="st">"lin"</span>,
                          <span class="kw">noInterpolAtEnds</span> <span class="kw">=</span> <span class="fl">TRUE</span>)
<span class="no">resM</span></pre></body></html></div>
<pre><code>## [1] 13.13147 15.54338 21.29288 24.60916  0.00000</code></pre>
<p>Plotting the results helps to better compare the differences. The black points in the figure below represent the actual values of <code>Y</code> and the grey vertical lines the breaks defining the bins. The blue lines and points represent the result from the <code>profBinLin</code> method. The bin values for the first and 4th bin are clearly wrong. The green colored points and lines represent the results from the <code>binYonX</code> and <code>imputeLinInterpol</code> functions (showing the correct binning and interpolation).</p>
<div class="sourceCode" id="cb11"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/plot.html">plot</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">X</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">Y</span>, <span class="kw">pch</span> <span class="kw">=</span> <span class="fl">16</span>, <span class="kw">ylim</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">Y</span>, <span class="kw">na.rm</span> <span class="kw">=</span> <span class="fl">TRUE</span>)),
     <span class="kw">xlim</span> <span class="kw">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">0</span>, <span class="fl">12</span>))
<span class="co">## Plot the breaks</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span>(<span class="kw">v</span> <span class="kw">=</span> <span class="fu"><a href="../reference/breaks_on_nBins.html">breaks_on_nBins</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">min</a></span>(<span class="no">X</span>), <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span>(<span class="no">X</span>), <span class="fl">5L</span>, <span class="fl">TRUE</span>), <span class="kw">col</span> <span class="kw">=</span> <span class="st">"grey"</span>)
<span class="co">## Result from profBinLin:</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">res</span>$<span class="no">x</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">resX</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="st">"blue"</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"b"</span>)
<span class="co">## Results from imputeLinInterpol</span>
<span class="fu"><a href="https://rdrr.io/r/graphics/points.html">points</a></span>(<span class="kw">x</span> <span class="kw">=</span> <span class="no">res</span>$<span class="no">x</span>, <span class="kw">y</span> <span class="kw">=</span> <span class="no">resM</span>, <span class="kw">col</span> <span class="kw">=</span> <span class="st">"green"</span>, <span class="kw">type</span> <span class="kw">=</span> <span class="st">"b"</span>,
       <span class="kw">pch</span> <span class="kw">=</span> <span class="fl">4</span>, <span class="kw">lty</span> <span class="kw">=</span> <span class="fl">2</span>)</pre></body></html></div>
<div class="figure" style="text-align: center">
<img src="new_functionality_files/figure-html/profBinLin-problems-1.png" alt="Illustration of the two bugs in profBinLin. The input values are represented by black points, grey vertical lines indicate the bins. The results from binning and interpolation with profBinLin are shown in blue and those from binYonX in combination with imputeLinInterpol in green." width="960"><p class="caption">
Illustration of the two bugs in profBinLin. The input values are represented by black points, grey vertical lines indicate the bins. The results from binning and interpolation with profBinLin are shown in blue and those from binYonX in combination with imputeLinInterpol in green.
</p>
</div>
<p>Note that by default <code>imputeLinInterpol</code> would also interpolate missing values at the beginning and the end of the provided numeric vector. This can be disabled (to be compliant with <code>profBinLin</code>) by setting parameter <code>noInterpolAtEnds</code> to <code>TRUE</code> (like in the example above).</p>
</div>
<div id="differences-due-to-updates-in-do_findchrompeaks_matchedfilter-respectively-findpeaks-matchedfilter-" class="section level2">
<h2 class="hasAnchor">
<a href="#differences-due-to-updates-in-do_findchrompeaks_matchedfilter-respectively-findpeaks-matchedfilter-" class="anchor"></a>Differences due to updates in <code>do_findChromPeaks_matchedFilter</code>, respectively <code>findPeaks.matchedFilter</code>.</h2>
<p>The original <code>findPeaks.matchedFilter</code> (up to version 1.49.7) had several shortcomings and bugs that have been fixed in the new <code>do_findChromPeaks_matchedFilter</code> method:</p>
<ul>
<li><p>The internal iterative processing of smaller chunks of the full data (also referred to as <em>iterative buffering</em>) could result, for some bin (step) sizes to unstable binning results (discussed in issue <a href="https://github.com/sneumann/xcms/issues/47">#47</a> on github): calculation of the breaks, or to be precise, the actually used bin size was performed in each iteration and could lead to slightly different sizes between iterations (due to rounding errors caused by floating point number representations in C).</p></li>
<li><p>The iterative buffering raises also a conceptual issue when linear interpolation is performed to impute missing values: the linear imputation will only consider values within the actually processed buffer and can thus lead to wrong or inaccurate imputations.</p></li>
<li><p>The <code>profBinLin</code> implementation contains two bugs, one that can result in failing to identify the maximal value in the first and last bin (see issue <a href="https://github.com/sneumann/xcms/issues/46">#46</a>) and one that fails to assign a value to a bin (issue <a href="https://github.com/sneumann/xcms/issues/49">#49</a>). Both are fixed in the <code>do_findChromPeaks_matchedFilter</code> implementation.</p></li>
</ul>
<p>A detailed description of tests comparing all implementations is available in issue <a href="https://github.com/sneumann/xcms/issues/52">#52</a> on github. Note also that in course of these changes also the <code>getEIC</code> method has been updated to use the new binning and missing value imputation function.</p>
<p>While it is strongly discouraged, it is still possible to use to <em>old</em> code (from 1.49.7) by calling <code><a href="../reference/useOriginalCode.html">useOriginalCode(TRUE)</a></code>.</p>
</div>
<div id="differences-in-findpeaks-massifquant" class="section level2">
<h2 class="hasAnchor">
<a href="#differences-in-findpeaks-massifquant" class="anchor"></a>Differences in <code>findPeaks.massifquant</code>
</h2>
<ul>
<li>Argument <code>scanrange</code> was ignored in the <em>original</em> old code (issue <a href="https://github.com/sneumann/xcms/issues/61">#61</a>).</li>
<li>The method returned a <code>matrix</code> if <code>withWave</code> was <code>0</code> and a <code>xcmsPeaks</code> object otherwise. The updated version returns <strong>always</strong> an <code>xcmsPeaks</code> object (issue #60).</li>
</ul>
</div>
<div id="differences-in-obiwarp-retention-time-correction" class="section level2">
<h2 class="hasAnchor">
<a href="#differences-in-obiwarp-retention-time-correction" class="anchor"></a>Differences in <em>obiwarp</em> retention time correction</h2>
<p>Retention time correction using the obiwarp method uses the <em>profile</em> matrix (i.e. intensities binned in discrete bins along the mz axis). Profile matrix generation uses now the <code>binYonX</code> method which fixed some problems in the original binning and linear interpolation methods. Thus results might be slightly different.</p>
<p>Also, the <code>retcor.obiwarp</code> method reports (un-rounded) adjusted retention times, but adjusts the retention time of eventually already identified peaks using rounded adjusted retention times. The new <code>adjustRtime</code> method(s) does adjust identified peaks using the reported adjusted retention times (not rounded). This guarantees that e.g. removing retention time adjustment/alignment results from an object restores the object to its initial state (i.e. the adjusted retention times of the identified peaks are reverted to the retention times before alignment). See issue <a href="https://github.com/sneumann/xcms/issues/122">#122</a> for more details.</p>
</div>
<div id="retcor-peaksgroups-change-in-the-way-how-well-behaved-peak-groups-are-ordered" class="section level2">
<h2 class="hasAnchor">
<a href="#retcor-peaksgroups-change-in-the-way-how-well-behaved-peak-groups-are-ordered" class="anchor"></a><code>retcor.peaksgroups</code>: change in the way how <em>well behaved</em> peak groups are ordered</h2>
<p>The <code>retcor.peakgroups</code> defines first the chromatographic peak groups that are used for the alignment of all spectra. Once these are identified, the retention time of the peak with the highest intensity in a sample for a given peak group is returned and the peak groups are ordered increasingly by retention time (which is required for the later fitting of either a polynomial or a linear model to the data). The selection of the retention time of the peak with the highest intensity within a feature (peak group) and samples, denoted as <em>representative</em> peak for a given feature in a sample, ensures that only the retention time of a single peak per sample and feature is selected (note that multiple chromatographic peaks within the same sample can be assigned to a feature). In the original code the ordering of the peak groups was however performed using the median retention time of the complete peak group (which includes also potential additional peaks per sample). This has been changed and the features are ordered now by the median retention time across samples of the representative chromatographic peaks.</p>
</div>
<div id="scanrange-parameter-in-all-findpeaks-methods" class="section level2">
<h2 class="hasAnchor">
<a href="#scanrange-parameter-in-all-findpeaks-methods" class="anchor"></a><code>scanrange</code> parameter in all <code>findPeaks</code> methods</h2>
<p>The <code>scanrange</code> in the <code>findPeaks</code> methods is supposed to enable the peak detection only within a user-defined range of scans. This was however not performed in each method. Due to a bug in <code>findPeaks.matchedFilter</code>’s original code the argument was ignored, except if the upper scan number of the user defined range was larger than the total number of available scans (see issue <a href="https://github.com/sneumann/xcms/issues/63">#63</a>). In <code>findPeaks.massifquant</code> the argument was completely ignored (see issue <a href="https://github.com/sneumann/xcms/issues/61">#61</a>) and, while the argument was considered in <code>findPeaks.centWave</code> and feature detection was performed within the specified scan range, but the original <code>@scantime</code> slot was used throughout the code instead of just the scan times for the specified scan indices (see issue <a href="https://github.com/sneumann/xcms/issues/64">#64</a>).</p>
<p>These problems have been fixed in version 1.51.1 by first sub-setting the <code>xcmsRaw</code> object (using the <code><a href="https://rdrr.io/r/base/Extract.html">[</a></code> method) before actually performing the feature detection.</p>
</div>
<div id="fillpeaks-fillchrompeaks-differences" class="section level2">
<h2 class="hasAnchor">
<a href="#fillpeaks-fillchrompeaks-differences" class="anchor"></a><code>fillPeaks</code> (<code>fillChromPeaks</code>) differences</h2>
<p>In the original <code>fillPeaks.MSW</code>, the mz range from which the signal is to be integrated was defined using</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="no">mzarea</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/which.min.html">which.min</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">mzs</span> - <span class="no">peakArea</span>[<span class="no">i</span>, <span class="st">"mzmin"</span>])),
          <span class="fu"><a href="https://rdrr.io/r/base/which.min.html">which.min</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="no">mzs</span> - <span class="no">peakArea</span>[<span class="no">i</span>, <span class="st">"mzmax"</span>])))</pre></body></html></div>
<p>Depending on the data this could lead to the inclusion of signal in the integration that are just outside of the mz range. In the new <code>fillChromPeaks</code> method signal is integrated only for mz values &gt;= mzmin and &lt;= mzmax thus ensuring that only signal is used that is truly within the peak area defined by columns <code>"mzmin"</code>, <code>"mzmax"</code>, <code>"rtmin"</code> and <code>"rtmax"</code>.</p>
<p>Also, the <code>fillPeaks.chrom</code> method did return <code>"into"</code> and <code>"maxo"</code> values of <code>0</code> if no signal was found in the peak area. The new method does not integrate any signal in such cases and does not fill in that peak.</p>
<p>See also issue <a href="https://github.com/sneumann/xcms/issues/130">#130</a> for more information.</p>
</div>
</div>
<div id="under-the-hood-changes" class="section level1">
<h1 class="hasAnchor">
<a href="#under-the-hood-changes" class="anchor"></a>Under the hood changes</h1>
<p>These changes and updates will not have any large impact on the day-to-day use of <code>xcms</code> and are listed here for completeness.</p>
<ul>
<li>From <code>xcms</code> version 1.51.1 on the default methods from the <code>mzR</code> package are used for data import. Besides ensuring easier maintenance, this enables also data import from <em>gzipped</em> mzML files.</li>
</ul>
</div>
<div id="deprecated-functions-and-files" class="section level1">
<h1 class="hasAnchor">
<a href="#deprecated-functions-and-files" class="anchor"></a>Deprecated functions and files</h1>
<p>Here we list all of the functions and related files that are deprecated.</p>
<ul>
<li><p><code>xcmsParallelSetup</code>, <code>xcmsPapply</code>, <code>xcmsClusterApply</code>: use <code>BiocParallel</code> package instead to setup and perform parallel processing, either <em>via</em> the <code>BPPARAM</code> parameter to function and methods, or by calling <code>register</code> to globally set parallel processing.</p></li>
<li><p><code>profBin</code>, <code>profBinM</code>, <code>profBinLin</code>, <code>profBinLinM</code>, <code>profBinLinBase</code>, <code>profBinLinBaseM</code>: replaced by the <code>binYonX</code> and <code>imputeLinInterpol</code> functions. Also, to create or extract the profile matrix from an <code>xcmsRaw</code> object, the <code>profMat</code> method.</p></li>
</ul>
<div id="deprecated" class="section level2">
<h2 class="hasAnchor">
<a href="#deprecated" class="anchor"></a>Deprecated</h2>
<div id="xcms-1-49" class="section level3">
<h3 class="hasAnchor">
<a href="#xcms-1-49" class="anchor"></a>xcms 1.49:</h3>
<ul>
<li>
<code>xcmsParallelSetup</code> (Deprecated.R)</li>
<li>
<code>xcmsPapply</code> (Deprecated.R)</li>
<li>
<code>xcmsClusterApply</code> (Deprecated.R)</li>
</ul>
</div>
<div id="xcms-1-51" class="section level3">
<h3 class="hasAnchor">
<a href="#xcms-1-51" class="anchor"></a>xcms 1.51:</h3>
<ul>
<li>
<code>profBin</code> (c.R)</li>
<li>
<code>profBinM</code> (c.R)</li>
<li>
<code>profBinLin</code> (c.R)</li>
<li>
<code>profBinLinM</code> (c.R)</li>
<li>
<code>profBinLinBase</code> (c.R)</li>
<li>
<code>profBinLinBaseM</code> (c.R)</li>
</ul>
</div>
</div>
<div id="defunct" class="section level2">
<h2 class="hasAnchor">
<a href="#defunct" class="anchor"></a>Defunct</h2>
</div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Tautenhahn:2008fx">
<p>1. Tautenhahn R, Böttcher C, Neumann S: <strong>Highly sensitive feature detection for high resolution LC/MS.</strong> <em>BMC Bioinformatics</em> 2008, <strong>9</strong>:504.</p>
</div>
<div id="ref-Smith:2006ic">
<p>2. Smith CA, Want EJ, O’Maille G, Abagyan R, Siuzdak G: <strong>XCMS: processing mass spectrometry data for metabolite profiling using nonlinear peak alignment, matching, and identification.</strong> <em>Analytical chemistry</em> 2006, <strong>78</strong>:779–787.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Steffen Neumann.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
